File: ptaas_fr/lib/models_2.dart
import 'package:json_annotation/json_annotation.dart';

part 'models_2.g.dart';

// dart run build_runner build

@JsonSerializable()
class Project {
  final String id;
  final bool installed;
  final List<Script> scripts;

  Project({required this.id, required this.installed, required this.scripts});

  factory Project.fromJson(Map<String, dynamic> json) =>
      _$ProjectFromJson(json);

  Map<String, dynamic> toJson() => _$ProjectToJson(this);
}

@JsonSerializable()
class Script {
  final String id;

  Script({required this.id});

  factory Script.fromJson(Map<String, dynamic> json) => _$ScriptFromJson(json);

  Map<String, dynamic> toJson() => _$ScriptToJson(this);
}

@JsonSerializable()
class APIError {
  final String message;
  final String? reason;

  APIError({required this.message, required this.reason});

  factory APIError.fromJson(Map<String, dynamic> json) =>
      _$APIErrorFromJson(json);

  Map<String, dynamic> toJson() => _$APIErrorToJson(this);
}

// Responses

@JsonSerializable()
class APIResponse {
  APIResponseProcessd? processed;
  APIResponseFailed? failed;

  APIResponse({this.processed, this.failed});

  factory APIResponse.fromJson(Map<String, dynamic> json) =>
      _$APIResponseFromJson(json);

  Map<String, dynamic> toJson() => _$APIResponseToJson(this);
}

@JsonSerializable()
class APIResponseProcessd {
  AllProjectsResponse? allProjects;
  AllScriptsResponse? allScripts;

  APIResponseProcessd({this.allProjects, this.allScripts});

  factory APIResponseProcessd.fromJson(Map<String, dynamic> json) =>
      _$APIResponseProcessdFromJson(json);

  Map<String, dynamic> toJson() => _$APIResponseProcessdToJson(this);
}

@JsonSerializable()
class APIResponseFailed {
  APIError? missingToken;
  APIError? emptyToken;
  APIError? notLoggedIn;
  APIError? internalServerError;

  APIResponseFailed(
      {this.missingToken,
      this.emptyToken,
      this.notLoggedIn,
      this.internalServerError});

  factory APIResponseFailed.fromJson(Map<String, dynamic> json) =>
      _$APIResponseFailedFromJson(json);

  Map<String, dynamic> toJson() => _$APIResponseFailedToJson(this);
}

// Projects

@JsonSerializable()
class AllProjectsResponse {
  AllProjectsResponseProcessed? processed;
  AllProjectsResponseFailed? failed;

  AllProjectsResponse({this.processed, this.failed});

  factory AllProjectsResponse.fromJson(Map<String, dynamic> json) =>
      _$AllProjectsResponseFromJson(json);

  Map<String, dynamic> toJson() => _$AllProjectsResponseToJson(this);
}

@JsonSerializable()
class AllProjectsResponseProcessed {
  List<Project> projects;

  AllProjectsResponseProcessed({required this.projects});

  factory AllProjectsResponseProcessed.fromJson(Map<String, dynamic> json) =>
      _$AllProjectsResponseProcessedFromJson(json);

  Map<String, dynamic> toJson() => _$AllProjectsResponseProcessedToJson(this);
}

@JsonSerializable()
class AllProjectsResponseFailed {
  APIError? cantReadProjects;
  APIError? aProjectIsMissing;

  AllProjectsResponseFailed({this.cantReadProjects, this.aProjectIsMissing});

  factory AllProjectsResponseFailed.fromJson(Map<String, dynamic> json) =>
      _$AllProjectsResponseFailedFromJson(json);

  Map<String, dynamic> toJson() => _$AllProjectsResponseFailedToJson(this);
}

// Scripts

@JsonSerializable()
class AllScriptsResponse {
  AllScriptsResponseProcessed? processed;
  AllScriptsResponseFailed? failed;

  AllScriptsResponse({this.processed, this.failed});

  factory AllScriptsResponse.fromJson(Map<String, dynamic> json) =>
      _$AllScriptsResponseFromJson(json);

  Map<String, dynamic> toJson() => _$AllScriptsResponseToJson(this);
}

@JsonSerializable()
class AllScriptsResponseProcessed {
  List<Script> scripts;

  AllScriptsResponseProcessed({required this.scripts});

  factory AllScriptsResponseProcessed.fromJson(Map<String, dynamic> json) =>
      _$AllScriptsResponseProcessedFromJson(json);

  Map<String, dynamic> toJson() => _$AllScriptsResponseProcessedToJson(this);
}

@JsonSerializable()
class AllScriptsResponseFailed {
  APIError? cantReadScripts;
  APIError? aScriptIsMissing;

  AllScriptsResponseFailed({this.cantReadScripts, this.aScriptIsMissing});

  factory AllScriptsResponseFailed.fromJson(Map<String, dynamic> json) =>
      _$AllScriptsResponseFailedFromJson(json);

  Map<String, dynamic> toJson() => _$AllScriptsResponseFailedToJson(this);
}

File: ptaas_rs/src/models_2.rs
use convertible::macros::DartConvertible;
use serde::{Deserialize, Serialize};
// Models

#[derive(Serialize, Deserialize, Debug, Clone)] //,DartConvertible)]
#[serde(rename_all = "camelCase")]
// #[dart_convertible(rename_all = "camelCase")]
pub struct Project {
    pub id: String,
    pub installed: bool,
    pub scripts: Vec<Script>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct Script {
    pub id: String,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct APIError {
    pub message: String,
    pub reason: String,
}

// Responses

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub enum APIResponse {
    Processed(APIResponseProcessd),
    Failed(APIResponseFailed),
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub enum APIResponseProcessd {
    AllProjects(AllProjectsResponse),
    AllScripts(AllScriptsResponse),
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub enum APIResponseFailed {
    MissingToken(APIError),
    EmptyToken(APIError),
    NotLoggedIn(APIError),
    InternalServerError(APIError),
}

// Projects

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub enum AllProjectsResponse {
    Processed(AllProjectsResponseProcessed),
    Failed(AllProjectsResponseFailed),
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct AllProjectsResponseProcessed {
    pub projects: Vec<Project>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub enum AllProjectsResponseFailed {
    CantReadProjects(APIError),
    AProjectIsMissing(APIError),
}

// Scripts

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub enum AllScriptsResponse {
    Processed(AllScriptsResponseProcessed),
    Failed(AllScriptsResponseFailed),
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct AllScriptsResponseProcessed {
    pub scripts: Vec<Script>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub enum AllScriptsResponseFailed {
    CantReadScripts(APIError),
    AScriptIsMissing(APIError),
}

pub fn print_dummies() {
    let api_failed = APIResponse::Failed(APIResponseFailed::MissingToken(APIError {
        message: "where the fuck is the token?".to_string(),
        reason: "permissions".to_string(),
    }));

    let all_proj = APIResponse::Processed(APIResponseProcessd::AllProjects(
        AllProjectsResponse::Processed(AllProjectsResponseProcessed {
            projects: vec![Project {
                id: "id".to_string(),
                installed: true,
                scripts: vec![Script {
                    id: "id".to_string(),
                }],
            }],
        }),
    ));

    let all_proj_failed = APIResponse::Processed(APIResponseProcessd::AllProjects(
        AllProjectsResponse::Failed(AllProjectsResponseFailed::AProjectIsMissing(APIError {
            message: "We are missing something".to_string(),
            reason: "permissions".to_string(),
        })),
    ));

    let all_scripts = APIResponse::Processed(APIResponseProcessd::AllScripts(
        AllScriptsResponse::Processed(AllScriptsResponseProcessed {
            scripts: vec![Script {
                id: "id".to_string(),
            }],
        }),
    ));

    let all_scripts_failed = APIResponse::Processed(APIResponseProcessd::AllScripts(
        AllScriptsResponse::Failed(AllScriptsResponseFailed::AScriptIsMissing(APIError {
            message: "Well that did not work".to_string(),
            reason: "permissions".to_string(),
        })),
    ));

    // print them with serde_json

    println!(
        "api_failed:\n{}\n",
        serde_json::to_string(&api_failed).unwrap()
    );
    println!("all_proj:\n{}\n", serde_json::to_string(&all_proj).unwrap());
    println!(
        "all_proj_failed:\n{}\n",
        serde_json::to_string(&all_proj_failed).unwrap()
    );
    println!(
        "all_scripts:\n{}\n",
        serde_json::to_string(&all_scripts).unwrap()
    );
    println!(
        "all_scripts_failed:\n{}\n",
        serde_json::to_string(&all_scripts_failed).unwrap()
    );
}
